### DISCUSSION TITLE
How do the derivatives computed from DAFoam find application in SNOPT?

### DISCUSSION THREAD

**Discussion by uttamcadambi07** [mdolab/dafoam#540] [General]
I guess this is a general (and perhaps very basic) question but something that I’ve been trying to think about. 

SNOPT uses SQP to do non linear optimisation. Here, it needs grad(f) and grad(c). These are partial derivatives of objectives and constraints. 

But DAFoam as we know supplies total derivatives with respect to the design variables (from the derivation using the adjoint equations). 

How is this derivative information used in SNOPT? I guess even for IPOPT I have this equation because I looked at the equations in IPOPT (the KKT conditions) and they don’t require total derivatives (I may be wrong). I was always under the impression that you will use a total derivative only for a gradient descent (but these non linear optimiser obviously don’t do that because we know that that is inefficient). What am I missing here?

**Comment by friedenhe** [comment]
The KKT condition does include the gradients for the obj and constraint functions. Check Chapter 5 of this book: https://mdobook.github.io/
